galsim/phase_psf.py

```diff
@@ -0,0 +1,1017 @@
+#
+"""@file phase_psf.py
+Utilities for creating PSFs from phase screens.  Essentially evaluates the Fourier optics
+diffraction equation:
+
+PSF(x, y) = int( |FT(aperture(u, v) * exp(i * phase(u, v, x, y, t)))|^2, dt)
+
+where x, y are focal plane coordinates and u, v are pupil plane coordinates
+
+The main classes of note are:
+
+FrozenAtmosphericScreen
+  Class implementing phase(u, v, x, y, t) for frozen flow von Karman type turbulence
+
+ARAtmosphericScreen
+  Class implementing phase(u, v, x, y, t) for "boiling" von Karman type turbulence
+
+PhaseScreenList
+  Python sequence type to hold multiple phase screens, for instance to simulate turbulence at
+  different altitudes.  A key method is getPSF(), which will take the list of phase screens,
+  add them together linearly (Fraunhofer approximation), and evaluate the above diffraction
+  equation to yield a PhaseScreenPSF object.
+
+PhaseScreenPSF
+  A GSObject holding the evaluated PSF.
+
+Atmosphere
+  Convenience function to quickly assemble multiple AtmosphericScreens into a PhaseScreenList.
+"""
+
+import copy
+import numpy as np
+import galsim
+import utilities
+from galsim import GSObject
+
+
+def listify(arg):
+    """Turn argument into a list if not already iterable."""
+    return [arg] if not hasattr(arg, '__iter__') else arg
+
+
+def broadcast(arg, n):
+    """Turn length-1 iterable into length-n list."""
+    return [arg[0]]*n if len(arg) == 1 else arg
+
+
+def generate_pupil(nu, pupil_size, diam=None, obscuration=None):
+    """ Generate a pupil transmission array (0's and 1's) for a circular aperture and potentially a
+    central circular obscuration.
+
+    @param nu          Number of pixels across pupil array.
+    @param pupil_size  Physical size of pupil array in meters.
+    @param diam        Diameter of aperture in meters.
+    @param obscuration Fractional linear obscuration.
+    @returns array of 0's and 1's indicating pupil transmission function.
+    """
+    aper = np.ones((nu, nu), dtype=np.float64)
+    if diam is not None:
+        radius = 0.5*diam
+        u = np.fft.fftshift(np.fft.fftfreq(nu, 1./pupil_size))
+        u, v = np.meshgrid(u, u)
+        rsqr = u**2 + v**2
+        aper = rsqr <= radius**2
+        if obscuration is not None:
+            aper *= rsqr >= (radius*obscuration)**2
+    return aper
+
+
+class PhaseScreen(object):
+    # ABC for phase screens.  Subclasses should implement:
+    # advance()
+    # advance_by()
+    # reset()
+    # path_difference()
+    # pupil_scale()
+    # Non-evolving screens, such as one representing optical aberrations, can probably accept the
+    # default advance(), advance_by(), and reset() methods which are all no-ops.  All subclasses
+    # will need to implement their own path_difference() and pupil_scale() though.
+    """ Abstract base class for a phase screen to use in generating a PSF using Fourier optics.
+    Not intended to be instantiated directly.
```

Comment:
>Is there a plan to extend this in a different direction so eventually there will be another subclass of this, where this abstraction might be relevant?

My thought was that a future OpticalPhaseScreen would also inherit from this, particularly a refactorization of @mdschneider's https://github.com/DarkEnergyScienceCollaboration/psf_model repo for field angle dependent LSST optical PSFs.